main.nf

```diff
@@ -46,46 +46,120 @@ params.sortThreads = 4
 /** Produce more output for debugging. */
 params.debug = false
 
+/** The Nextflow publish mode. See https://www.nextflow.io/docs/latest/process.html#publishdir */
+params.publishMode = "rellink"
 
 
+/** Parameter checking and casting code */
 allowedParameters = ['input', 'outputDir',
                      'sortFastqs', 'sortMemory', 'sortThreads',
                      'compressIntermediateFastqs', 'compressorThreads',
                      'excludedFlags', 'checkIntermediateFastqMd5',
-                     'debug']
+                     'debug', 'publishMode']
 
 checkParameters(params, allowedParameters)
 
 // The sorting memory is used as MemoryUnit in the code below.
 sortMemory = new MemoryUnit(params.sortMemory)
 
+/** See https://www.nextflow.io/docs/latest/process.html#publishdir */
+enum PublishMode {
+    symlink, // Absolute symbolic link in the published directory for each process output file.
+    rellink, // Relative symbolic link in the published directory for each process output file.
+    link,    // Hard link in the published directory for each process output file.
+    copy,    // Copies the output files into the published directory.
+    copyNoFollow, // Copies the output files into the published directory without following
+                  // symlinks ie. copies the links themselves.
+    move;     // Moves the output files into the published directory. Note: this is only supposed
+              // to be used for a terminating process i.e. a process whose output is not consumed
+              // by any other downstream process.
+
+    static PublishMode fromString(String str) throws IllegalArgumentException {
+        return str.toLowerCase() as PublishMode
+    }
+}
+
+publishMode = PublishMode.fromString(params.publishMode)
+
+
+log.info """
+==================================
+= nf-bam2fastq                   =
+==================================
+${allowedParameters.collect { "$it = ${params.get(it)}" }.join("\n")}
 
-def fastqSuffix(Boolean compressed) {
+"""
+
+
+String fastqSuffix(Boolean compressed) {
     if (compressed)
         return "fastq.gz"
     else
         return "fastq"
 }
 
 
-def sortedFastqFile(String outDir, Path unsortedFastq, Boolean compressed) {
+String sortedFastqFile(String outDir, Path unsortedFastq, Boolean compressed) {
     String result = outDir + "/" + unsortedFastq.getFileName().toString().replaceFirst(/\.fastq(?:\.gz)?$/, ".sorted.fastq")
     if (compressed)
         return result + ".gz"
     else
         return result
 }
 
+/** Check whether parameters are correct (names and values)
```

Comment:
Moved parameter checking code upwards. Seemed lost at the end of the file. Also its strange to use functions that are just defined later in the same file.